internal/pally/registry.go

```
@@ -0,0 +1,250 @@
+package pally
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/prometheus/common/model"
+	"github.com/uber-go/tally"
+	"go.uber.org/atomic"
+)
+
+// A Registry is a collection of metrics, usually scoped to a single package or
+// object. Each Registry is also its own http.Handler, and can serve
+// Prometheus-flavored text and protocol buffer pages for metrics introspection
+// or scraping.
+type Registry struct {
+	http.Handler
+
+	metricsMu   sync.RWMutex
+	metrics     []metric
+	constLabels Labels
+
+	// TODO: To serve our own Prometheus endpoints, we only need to implement
+	// prometheus.Gatherer. The vanilla Prometheus registry is a convenient
+	// implementation, but it enforces metric uniqueness via FNV64a; if we run
+	// into hash collisions in our more-demanding applications, we should use a
+	// simple map[string]struct{} instead.
+	prom      *prometheus.Registry
+	federated []prometheus.Registerer
+
+	pushing atomic.Bool
+	scope   tally.Scope
+	ticker  *time.Ticker
+	stop    chan struct{}
+	stopped chan struct{}
+}
+
+// A RegistryOption configures a Registry.
+type RegistryOption func(*Registry)
+
+// Federated links a pally.Registry with a prometheus.Registerer, so that all
+// metrics created in one also appear in the other.
+func Federated(prom prometheus.Registerer) RegistryOption {
+	return func(r *Registry) {
+		r.federated = append(r.federated, prom)
+	}
+}
+
+// Labeled adds constant labels to a Registry. All metrics created by a
+// Registry inherit its constant labels.
+func Labeled(ls Labels) RegistryOption {
+	return func(r *Registry) {
+		for k, v := range ls {
+			if !model.LabelName(k).IsValid() || !model.LabelValue(v).IsValid() {
+				continue
+			}
+			if !isValidTallyString(k) || !isValidTallyString(v) {
+				continue
```

This will silently skip invalid labels. Are we okay with that?